src/GovUk.Education.ExploreEducationStatistics.Data.Api.Tests/Controllers/FastTrackControllerTests.cs

```diff
@@ -1,86 +1,316 @@
+                .ReturnsAsync(null);
+
             mocks
                 .fastTrackService
                 .Setup(s => s.GetFastTrackAndResults(fastTrackId))
                 .ReturnsAsync(fastTrackViewModel);
 
+            mocks.cacheService
+                .Setup(s => s.SetItem<object>(cacheKey, fastTrackViewModel))
+                .Returns(Task.CompletedTask);
+
+            var result = await controller.Get(fastTrackId.ToString());
+            
+            VerifyAllMocks(mocks);
+
+            result.AssertOkResult(fastTrackViewModel);
+        }
+
+        [Fact]
+        public async Task Get_NotFound()
+        {
+            var fastTrackId = Guid.NewGuid();
+            var cacheKey = new FastTrackResultsCacheKey("publication", "release", fastTrackId);
+
+            var (controller, mocks) = BuildControllerAndMocks();
+
             mocks
                 .cacheKeyService
                 .Setup(s => s.CreateCacheKeyForFastTrackResults(fastTrackId))
-                .ReturnsAsync(new FastTrackResultsCacheKey("publication", "release", fastTrackId));
+                .ReturnsAsync(cacheKey);
             
+            mocks.cacheService
+                .Setup(s => s.GetItem(cacheKey, typeof(FastTrackViewModel)))
+                .ReturnsAsync(null);
+
+            mocks
+                .fastTrackService
+                .Setup(s => s.GetFastTrackAndResults(fastTrackId))
+                .ReturnsAsync(new NotFoundResult());
+
             var result = await controller.Get(fastTrackId.ToString());
             VerifyAllMocks(mocks);
             
-            result.AssertOkResult(fastTrackViewModel);
+            result.AssertNotFoundResult();
         }
 
         [Fact]
-        public async Task Get_NotFound()
+        public async Task Get_NotFoundCreatingCacheKey()
         {
+            var fastTrackId = Guid.NewGuid();
+            
             var (controller, mocks) = BuildControllerAndMocks();
             
             mocks
-                .fastTrackService
-                .Setup(s => s.GetFastTrackAndResults(fastTrackId))
+                .cacheKeyService
+                .Setup(s => s.CreateCacheKeyForFastTrackResults(fastTrackId))
                 .ReturnsAsync(new NotFoundResult());
 
-            var result = await controller.Get("InvalidGuid");
+            var result = await controller.Get(fastTrackId.ToString());
+            VerifyAllMocks(mocks);
+            
             result.AssertNotFoundResult();
         }
 
         [Fact]
         public async Task Get_InvalidId()
         {
-            var (controller, mocks) = BuildControllerAndMocks();
+            var (controller, _) = BuildControllerAndMocks();
             var result = await controller.Get("InvalidGuid");
             result.AssertNotFoundResult();
         }
-        
-        private (
+
+        [Fact]
+        public void FastTrackViewModel_SerialiseAndDeserialise()
```

Comment:
The reason behind this is that we've had 2 issues whereby caching has broken but we've not been aware of it until it was revealed with manual testing / checking the log messages.

One time was due to something having an `init` rather than a `set` setter on a property, meaning that it could be written wholly to blob storage, but that particular field could not be retrieved as it had to setter.

The second was due to it not knowing how to deserialise the Location hierarchy / non-hierarchy property of a view model because it only had an `ISubjectMetViewModel` interface to work with when deserialising.

I'd hope that these would help us with such cases in the future.

Reply:
Good to know thanks. This could be relevant to the way I'm testing some changes around Permalinks. 🙂